Fixed Frame Option & Add Support for Personal Space Albums

homer05 opened the pull request:
# Add Fixed Frame Option for Slideshow Display

## Summary

This PR adds a new feature that allows the slideshow to display images within a fixed frame instead of fullscreen. This provides flexibility for users who want to position the slideshow in a specific area of their MagicMirror screen, leaving space for other modules.

## New Configuration Options

### `useFixedFrame` (Boolean)
- Enables or disables the fixed frame mode
- **Default**: `false`
- **Example**: `useFixedFrame: true`

### `frameWidth` (String)
- Width of the fixed frame (accepts CSS values like percentages or pixels)
- **Default**: `'80%'`
- **Example**: `frameWidth: '70%'` or `frameWidth: '1200px'`

### `frameHeight` (String)
- Height of the fixed frame (accepts CSS values like percentages or pixels)
- **Default**: `'80%'`
- **Example**: `frameHeight: '70%'` or `frameHeight: '800px'`

### `framePosition` (String)
- Position of the fixed frame on the screen
- **Default**: `'center'`
- **Valid Values**: `'center'`, `'top'`, `'bottom'`, `'left'`, `'right'`, `'top-left'`, `'top-right'`, `'bottom-left'`, `'bottom-right'`
- **Example**: `framePosition: 'top-right'`

### `frameBackgroundColor` (String)
- Background color for the area outside the frame (accepts any valid CSS color)
- **Default**: `'rgba(0, 0, 0, 0.5)'`
- **Example**: `frameBackgroundColor: 'rgba(0, 0, 0, 0.7)'` or `frameBackgroundColor: '#000000'`

## Usage Example

config.js
{
  module: 'MMM-SynPhotoSlideshow',
  config: {
    // ... existing configuration ...
    
    useFixedFrame: true,
    frameWidth: '80%',
    frameHeight: '80%',
    framePosition: 'center',
    frameBackgroundColor: 'rgba(0, 0, 0, 0.5)'
  }
}
```

## Technical Details

### Files Modified
- `src/types.ts` - Added new configuration properties to `ModuleConfig` interface
- `src/MMM-SynPhotoSlideshow.ts` - Added default values for new options
- `src/frontend/ModuleController.ts` - Added frame positioning and styling logic
- `src/frontend/display.scss` - Added CSS classes for fixed frame display
- `src/frontend/ConfigValidator.ts` - Added validation for new configuration options

### Features
- Dynamic frame sizing using CSS custom properties
- Flexible positioning with 9 position options
- Customizable background overlay
- Rounded corners and box shadow for visual appeal
- Fully backward compatible (defaults to fullscreen when disabled)

## Backward Compatibility

✅ **Fully backward compatible** - All new options have default values that maintain existing fullscreen behavior. Existing configurations will work without any changes.

Commit:
add option to resize slideshow

Commit:
Add support for personal space albums in Synology integration

homer05 commented:
# Add Support for Personal Space Albums in Synology Integration

## Summary

This PR adds support for KI-generated personal space albums in Synology Photos, allowing users to display photos from persons, concepts (themes), and geocoding (locations) albums.

## New Features

### Personal Space Album Support

The module now supports three types of KI-generated albums from Synology Photos personal space:

1. **Person Albums** - Photos tagged with specific persons (face recognition)
2. **Concept Albums** - Photos grouped by themes/concepts (e.g., cats, landscapes)
3. **Geocoding Albums** - Photos grouped by location

### New Configuration Options

#### `synologyPersonIds` (Array of numbers)
- IDs of person albums from personal space
- Example: `synologyPersonIds: [3503, 1804]`
- .env: `SYNOLOGY_PERSON_IDS=3503,1804`

#### `synologyConceptIds` (Array of numbers)
- IDs of concept/theme albums from personal space
- Example: `synologyConceptIds: [354]`
- .env: `SYNOLOGY_CONCEPT_IDS=354`

#### `synologyGeocodingIds` (Array of numbers)
- IDs of geocoding/location albums from personal space
- Example: `synologyGeocodingIds: [8]`
- .env: `SYNOLOGY_GEOCODING_IDS=8`

## Usage Example

```javascript
{
  module: 'MMM-SynPhotoSlideshow',
  config: {
    synologyPersonIds: [3503],      // Person albums
    synologyConceptIds: [354],      // Concept albums
    synologyGeocodingIds: [8]       // Geocoding albums
  }
}
```

## Technical Details

### Implementation
- Separate API calls for each album type using `person_id`, `concept_id`, and `geocoding_id` parameters
- Automatic retry with reduced limit on error 120 (limit condition)
- Priority: Personal space albums > Tags > Shared album > Regular albums

### Files Modified
- `src/types.ts` - Added new configuration properties
- `src/MMM-SynPhotoSlideshow.ts` - Added default values
- `src/backend/SynologyPhotosClient.ts` - Implemented album fetching logic
- `src/backend/SynologyManager.ts` - Added priority handling
- `src/backend/ConfigLoader.ts` - Added .env support

## Backward Compatibility

✅ **Fully backward compatible** - All new options have default empty arrays. Existing configurations continue to work without changes.

Commit:
Add logLevel configuration to filter DEBUG messages

Added logLevel option to ModuleConfig to control logging verbosity. DEBUG messages are now filtered based on the configured log level (default: 'info'). Supports 'error', 'warn', 'info', and 'debug' levels. Can be configured via config.js or LOG_LEVEL environment variable.

spyder007 commented on src/backend/ConfigLoader.ts:
```diff
@@ -122,6 +122,30 @@ class ConfigLoader {
     }
 
+    // Person IDs (comma-separated numbers)
+    // KI-generated person albums from personal_space
```
what does `KI-generated` mean?  it shows up in a few places, and I understand what this is (the albums that are auto-generated by the Photos app), but is KI the Synology specific AI client, or is this just a typo?

spyder007 commented on src/backend/ConfigLoader.ts:
```diff
@@ -122,6 +122,30 @@ class ConfigLoader {
+    }
+
+    // Concept IDs (comma-separated numbers)
+    // KI-generated concept albums (themes) from personal_space
```
Same as above

spyder007 commented on src/backend/ConfigLoader.ts:
```diff
@@ -122,6 +122,30 @@ class ConfigLoader {
+    }
+
+    // Geocoding IDs (comma-separated numbers)
+    // KI-generated geocoding albums (locations) from personal_space
```
Same as above

spyder007 commented on src/backend/SynologyPhotosClient.ts:
```diff
@@ -317,6 +326,183 @@ class SynologyPhotosClient {
   }
 
+  /**
+   * Fetch photos from a person album (KI-generated album)
```
Same as above

spyder007 commented on src/backend/SynologyPhotosClient.ts:
```diff
@@ -317,6 +326,183 @@ class SynologyPhotosClient {
+  }
+
+  /**
+   * Fetch photos from a concept album (KI-generated album)
```
Same as above

spyder007 commented on src/backend/SynologyPhotosClient.ts:
```diff
@@ -317,6 +326,183 @@ class SynologyPhotosClient {
+  }
+
+  /**
+   * Fetch photos from a geocoding album (KI-generated album)
```
Same as above

spyder007 commented on src/backend/SynologyManager.ts:
```diff
@@ -28,18 +28,34 @@ class SynologyManager {
       }
 
-      if (config.synologyTagNames && config.synologyTagNames.length > 0) {
+      // Personal space album IDs (persons, concepts, geocoding) have priority over tags and regular albums
```
Do these have to take precedence, or can they just all be included in the scraping?  Right now, the module is setup in such a way that, for tags, it pulls images with a tag from both the shared and personal spaces.

The end goal is simply a collection of photos to show, so I don't think we should preclude people from setting `synologyAlbumName` to grab all photos from an album, but also include personIds and conceptIds to pull images from those dynamically generated albums.

As I read this i realize that if you set both `synologyAlbumName` and `synologyTagNames`, the tags are preferred.  I added issue #13 to address this

spyder007 commented on src/backend/SynologyPhotosClient.ts:
You are missing new tests for this new functionality.  See `SynologyPhotosClient.test.ts`

spyder007 commented on src/frontend/ConfigValidator.ts:
Missing tests for this new functionality.  see `ConfigValidator.test.ts`

spyder007 commented on src/frontend/ModuleController.ts:
Missing tests.  see `ModuleController.test.ts`

spyder007 commented on src/types.ts:
```diff
@@ -16,6 +16,9 @@ export interface ModuleConfig {
   synologyPassword: string;
   synologyAlbumName: string;
   synologyTagNames: string[];
+  synologyPersonIds: number[];
+  synologyConceptIds: number[];
```
This comment from Copilot is fine, but also need to update `README.md` with your new settings and a good description.

spyder007 commented on src/types.ts:
```diff
@@ -55,6 +58,12 @@ export interface ModuleConfig {
   maxWidth: number;
   maxHeight: number;
   imageInfoNoFileExt: boolean;
+  useFixedFrame: boolean;
+  frameWidth: string;
+  frameHeight: string;
+  framePosition: string;
```
In code comments are nice, but make sure the README.md is updated.

spyder007 commented on src/backend/SynologyPhotosClient.ts:
```diff
@@ -317,6 +326,183 @@ class SynologyPhotosClient {
+
+  /**
+   * Fetch photos from persons, concepts, and geocoding albums
+   * These are KI-generated albums from the personal_space
```
Same as above

spyder007 commented on src/backend/Logger.ts:
These changes broke tests in `Logger.test.ts`

spyder007 commented on src/backend/ConfigLoader.ts:
Missing tests.  See `ConfigLoader.test.ts`

spyder007 reviewed (changes requested):
I love the changes, this will definitely increase functionality and be a nice to have for more.

A couple of things:

1. For new functionality, make sure you have unit test coverage.
2. With the broken tests in the Logger, you won't be able to see the sonar scan.  Once you fix the tests, a new build should generate a [Pull Request scan in SonarCloud](https://sonarcloud.io/project/pull_requests_list?id=spydersoft-consulting_MMM-SynPhotoSlideshow).  Make sure that any new issues are addressed.
3. Please make sure the documentation is updated with your changes.  At minimum, the README.md in the root, but give the extent of these changes, it might be nice to have additional files added to `docs` which detail the display frame and personal space albums support.

Commit:
Fix: Deaktiviere Frontend-Timer wenn Backend Slideshow steuert

- Entferne unused @ts-expect-error Direktive in ConfigLoader.ts
- Deaktiviere Frontend-Timer in ModuleController.ts (resume() startet keinen Timer mehr)
- Verhindere automatische NEXT_IMAGE Anfragen vom Frontend
- FÃ¼ge Logging hinzu um slideshowSpeed Wert zu verfolgen
- Backend steuert jetzt vollstÃ¤ndig die Slideshow-Timing basierend auf SLIDESHOW_SPEED

Fixes: Bilder wechseln jetzt korrekt alle 40 Sekunden statt alle 10 Sekunden

Commit:
Fix: Konsistente Cache-Keys und SYNOLOGY_MAX_PHOTOS aus .env

- Cache-Key-Generierung: Verwendung von synologyId und spaceId fÃ¼r konsistente Dateinamen (thumbnail_XXX_Y.jpg, original_XXX_Y.heic) statt Hash-basierter Keys
- ImageCache: getCacheKey() erkennt bereits generierte Keys und Synology-URLs
- ImageProcessor: Ãœbergabe von synologyId/spaceId an Cache-Methoden
- BackgroundDownloader: Ãœbergabe von synologyId/spaceId an Cache-Methoden
- SlideshowController: Ãœbergabe von synologyId/spaceId beim Preload
- ConfigLoader: SYNOLOGY_MAX_PHOTOS wird jetzt korrekt aus .env Datei geladen
- Debug-Logs hinzugefÃ¼gt fÃ¼r bessere Diagnose von Cache-Key-Generierung und Config-Loading
- PhotoStation Authentifizierung wiederhergestellt (optional, nicht kritisch)

Commit:
Implement dynamic batch loading: SYNOLOGY_MAX_PHOTOS as batch size, preload next batch when 2 images remain, remove periodic timers

Commit:
Always load metadata from original files, add centralized metadata database (location, captureDate) outside cache directory

Commit:
Implement reverse geocoding with address storage and optimize metadata loading

- Added reverse geocoding via Nominatim API with accept-language=en for Latin characters
- Store FullAddress and ShortAddress in centralized photo_metadata.json
- Load metadata from database first, only download original files if metadata missing
- Implement atomic file writing with backup mechanism to prevent corruption
- Add cache key limit handling to prevent maxKeys exceeded errors
- Reduce debug logging for EXIF metadata extraction
- Improve error handling for geocoding API calls

Commit:
Improve error handling for reverse geocoding and add response validation

- Enhanced error logging with detailed error classification (timeout, network, HTTP errors)
- Add response validation for empty responses and error fields
- Log full error objects in debug mode for better troubleshooting
- Add Accept header and validateStatus to axios request
- Improve handling of HTTP 4xx errors and rate limiting

Commit:
Add captureDate and FullAddress display from photo_metadata.json

- Add new imageInfo options: capturedate, fulladdress, shortaddress, address
- Display metadata from photo_metadata.json only (no EXIF fallback)
- Convert captureDate from UTC to CET/CEST timezone
- Position image info relative to image (inside transition div)
- Update ConfigValidator to accept new imageInfo values

Commit:
Add geolocation map and enhanced metadata display features

Features:
- Add interactive Leaflet map showing photo location above image info
- Display captureDate and FullAddress/ShortAddress from photo_metadata.json
- Automatic UTC to CET/CEST timezone conversion for dates
- New imageInfo options: capturedate, fulladdress, shortaddress, address
- Configurable map zoom level (mapZoom option)
- German map labels using tile.openstreetmap.de
- Metadata-only display (no EXIF fallback)

Technical changes:
- Add Leaflet.js library via CDN
- Parse location string from photo_metadata.json
- Position map relative to image (inside transition div)
- Update ConfigValidator to accept new imageInfo values
- Add comprehensive fork differences documentation (DE/EN)

This is a test version with all additional functionalities.

Commit:
Update FORK_DIFFERENCES.md with comprehensive feature documentation

Documentation now focuses on 5 main feature areas:
1. Fixed Frame Mode - Option for fullscreen vs. frame display
2. Advanced Filtering - Filter by persons, places (geocoding), or scenes (concepts)
3. Enhanced Image Cache & Metadata - Automatic original download and address determination
4. Enhanced Image Information Display - Display date and addresses from metadata
5. Geolocation Map - Interactive map with photo location

Includes detailed explanations in German and English for all features,
configuration examples, technical implementation details, and migration guide.

Commit:
Update README.md with new configuration options (DE/EN)

Added comprehensive documentation for new features:
- synologyPersonIds, synologyConceptIds, synologyGeocodingIds (AI-generated album filtering)
- useFixedFrame, frameWidth, frameHeight, framePosition, frameBackgroundColor (Fixed frame mode)
- Enhanced imageInfo options (capturedate, fulladdress, shortaddress, address, imagecount)
- mapZoom (geolocation map zoom level)

All options documented in both German and English with examples,
default values, and environment variable support where applicable.

Added new sections:
- Filtering by AI-Generated Albums
- Fixed Frame Mode
- Enhanced Image Information Display

Commit:
Add metadata database analysis and protect photo_metadata.json from deletion

Features:
- Add analyzeMetadataDatabase() method to track metadata statistics
- Display photos with location but no address (can be geocoded)
- Display photos without location (cannot be geocoded)
- Automatic analysis on startup and after metadata updates
- Log when new photos are added or addresses are updated

Protection:
- Explicitly protect photo_metadata.json from deletion in ImageCache.clear()
- Explicitly protect photo_metadata.json from deletion in ImageCache.evictOldFiles()
- Add documentation that photo_metadata.json must never be deleted
- File is stored outside cache directory for additional protection

The metadata file is now fully protected and will never be deleted, even when:
- Cache is cleared
- Old files are evicted
- New images are loaded from Synology
- filesShownTracker.txt is reset

Commit:
Add world map and fix positioning of image info and maps

Features:
- Add world map (left side, zoom level 2) showing location on world map
- Add detailed map (right side, configurable zoom via mapZoom)
- Fixed positioning: image info always left top, world map right top, detailed map right bottom
- Both maps have same dimensions (200x150px)
- Maps only displayed when location data is available from photo_metadata.json

Positioning:
- Image info: fixed left top (overrides imageInfoLocation setting)
- World map: fixed right top
- Detailed map: fixed right bottom (below world map with 10px gap)

The detailed map uses configurable zoom level from config.js (mapZoom option).

homer05 committed:
Merge pull request #1 from homer05/Redesign

Redesign

homer05 committed:
Update README.md

homer05 committed:
Add files via upload

homer05 committed:
Update README.md

Commit:
Fix Bugbot issues: logout log, memory leak, and deprecated option

Fixes:
1. Logout log message never displayed (SynologyPhotosClient.ts)
   - Store session state before nullification to properly log logout

2. Leaflet maps memory leak (UIBuilder.ts + ModuleController.ts)
   - Store map instances in container (_leafletMap property)
   - Add destroyMap() method to properly cleanup Leaflet instances
   - Call destroyMap() before removing map containers to prevent memory leaks

3. imageInfoLocation option deprecated (display.scss + README.md)
   - Document that imageInfoLocation no longer has any effect
   - Image info is now always positioned at top-left
   - Add CSS comments explaining deprecated classes
   - Update README.md with deprecation notice

These fixes address all issues identified by Cursor Bugbot review.

Commit:
Fix test failures after code changes

Test fixes:
- ImageCache.test.ts: Update tests for new return types (file paths instead of base64/boolean)
  - get() now returns file path instead of base64 data
  - set() now returns file path instead of boolean
  - Add keys() mock to NodeCache instance
  - Fix eviction test timeout

- ImageListManager.test.ts: Update for new getNextImage() behavior
  - getNextImage() now returns null at end instead of looping back
  - Update single image list test accordingly

- Logger.test.ts: Fix debug() tests
  - Set logLevel to 'debug' before testing debug() method
  - debug() only logs when logLevel is set to 'debug'

- UIBuilder.test.ts: Update date property tests
  - date property now only uses metadata.captureDate, not imageDate
  - Update tests to check for formatted date strings

Bugfixes:
- SlideshowController.ts: Fix preload logic to use newPhotos instead of refetching
- ModuleController.ts: Fix TypeScript error - change NodeJS.Timeout to number

These fixes address test failures identified after running npm run test.

Commit:
Fix memory leak: Destroy Leaflet maps before removing transition divs

Bug fix:
- Move map cleanup logic BEFORE cleanupOldImages() call
- Destroy Leaflet map instances from old transition divs before they are removed
- Prevents memory leaks from Leaflet map instances accumulating over time

The previous code searched for maps in the new empty transitionDiv, but the old maps
were in transition divs that were already removed by cleanupOldImages(). This caused
Leaflet maps with their event listeners and tile caches to remain in memory.

Now maps are properly destroyed from all old transition divs before cleanupOldImages()
removes them, preventing the memory leak.

Fixes issue identified by Cursor Bugbot review.

Commit:
Fix: Clear pending invalidateSize timers when destroying Leaflet maps

Bug fix:
- Store setTimeout timer reference in _leafletInvalidateTimer
- Clear timer in destroyMap() before removing map instance
- Add safety checks and try-catch around invalidateSize() call
- Prevents errors when invalidateSize() is called on destroyed maps

The previous code scheduled setTimeout to call map.invalidateSize() 50ms after
map creation, but if destroyMap() was called during fast image transitions,
the pending timer would still fire and try to call invalidateSize() on a
destroyed map, causing unhandled errors.

Now the timer is properly cleared when the map is destroyed, and additional
safety checks ensure no errors occur even if the timer somehow fires.

Fixes issue identified by Cursor Bugbot review.

homer05 committed:
Merge pull request #2 from homer05/Redesign

Redesign

Commit:
Add fork notice and link to FORK_DIFFERENCES.md at top of README

- Add prominent notice that this is a fork of the original repository
- Add link to docs/FORK_DIFFERENCES.md for complete list of differences
- Notice in both English and German
- Positioned at the very top for immediate visibility

Commit:
Add screenshot image

Commit:
Merge branch 'main' of https://github.com/homer05/MMM-SynPhotoSlideshow

homer05 commented:
i did a few other updates

homer05 commented:
I'm not sure we should really combine them https://github.com/homer05/MMM-SynPhotoSlideshow
